### PR TITLE
Add variables support in sql command

### PR DESCRIPTION
**Pull request by sfc-gh-turbaszek**
### Pre-review checklist
   * [ ] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [ ] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually.
   * [ ] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [ ] I've described my changes in the section below.

### Changes description
...

### CODE REVIEW

File: src/snowflake/cli/api/commands/flags.py [hunk 1]
```diff
@@ -485,7 +485,7 @@ def project_definition_option(project_name: str):
 
     def _callback(project_path: Optional[str]):
         dm = DefinitionManager(project_path)
-        project_definition = getattr(dm.project_definition, project_name, None)
+        project_definition = dm.project_definition
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/commands/flags.py:488]: This change alters what is provided by `@with_project_definition` decorator. Instead of part of project definition (snowpark, streamlit, native_app) it provides full definition.

The only purpose of keeping `@with_project_definition(NAME)` is to add `-p/--project` with "nice" description that relates to project type.

However, we already have some name mapping:
https://github.com/snowflakedb/snowflake-cli/blob/532d1f10a8be36105ec689ce597d20ac6bbd8e50/src/snowflake/cli/api/commands/flags.py#L500-L505

So my suggestion would be to get rid of this "nice description" thing in favor of simplicity. In this way we can move the functionality of this decorator to `app.command(uses_project=True)`.

@sfc-gh-cgorrie @sfc-gh-pjob wdyt?

File: src/snowflake/cli/api/project/schemas/project_definition.py [hunk 2]
```diff
@@ -1,19 +1,24 @@
...
 
 
+# todo: update examples
+_supported_version = ("1", "1.1")
+_latest_version = "1.1"
+
+
 class ProjectDefinition(UpdatableModel):
-    definition_version: int = Field(
+    definition_version: str = Field(
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/project/schemas/project_definition.py:19]: Had to change it from integer to string to enable minimal semantic versioning.

File: src/snowflake/cli/api/utils/rendering.py [hunk 3]
```diff
@@ -88,22 +94,151 @@ def jinja_render_from_file(
...
+            f"{context_key} in user defined data. The `{context_key}` variable is reserved for CLI usage."
+        )
+
+    project_definition = cli_context.project_definition
+    context_data = {context_key: {"env": _EnvGetter(os.environ)}}
+    # If there's project definition file then resolve variables from it
+    if project_definition and project_definition.meets_version_requirement("1.1"):
+        context_data = _resolve_variables_in_project(
+            project_definition=project_definition, context_key=context_key
+        )
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/utils/rendering.py:160]: This feature will work only for snowflake.yml with 1.1 or later version

File: src/snowflake/cli/api/commands/flags.py (hunk 1 above)
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/commands/flags.py:488]: should uses_project=True be the default?

File: src/snowflake/cli/api/utils/rendering.py [hunk 4]
```diff
@@ -88,22 +94,151 @@ def jinja_render_from_file(
         print(rendered_result)
 
 
-class _AttrGetter:
+def _is_key_in_environ(key: str):
+    return key.upper() in os.environ or key in os.environ
+
+
+def _get_value_from_environ(key: str):
+    if key.upper() in os.environ:
```
[the remark this answers is not in the document]
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/utils/rendering.py:102]: I can see we return key as a backup. I think it's safer to check key first, and then upper as a backup.

File: src/snowflake/cli/api/utils/rendering.py [hunk 5]
```diff
@@ -88,22 +94,151 @@ def jinja_render_from_file(
...
+        return os.environ[key.upper()]
+    return os.environ[key]
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/utils/rendering.py:104]: I think we should do key first, and if not found, try upper.

Comment (sfc-gh-cgorrie) [reply src/snowflake/cli/api/utils/rendering.py:104]: I'm not sure we should be trying to retrofit case-insensitivity onto a case-sensitive platform.

See https://docs.python.org/3/library/os.html#os.environ -- it already takes care of this on relevant platforms. So we can use `key` as-is.

File: src/snowflake/cli/api/utils/rendering.py [hunk 6]
```diff
@@ -88,22 +94,151 @@ def jinja_render_from_file(
         print(rendered_result)
 
 
-class _AttrGetter:
+def _is_key_in_environ(key: str):
+    return key.upper() in os.environ or key in os.environ
```
Comment (sfc-gh-cgorrie) [inline src/snowflake/cli/api/utils/rendering.py:98]:
```suggestion
    return key in os.environ
```
See below for rationale.

File: src/snowflake/cli/api/utils/rendering.py (hunk 5 above)
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/utils/rendering.py:104]: even better.

File: src/snowflake/cli/api/utils/rendering.py [hunk 7]
```diff
@@ -88,22 +94,151 @@ def jinja_render_from_file(
...
+
+class _ContextGetter:
+    """
+    Getter for project definition data. It replaces `env` attribute with _EnvGetter
+    that enables fallback to environment variables.
+    """
+
+    def __init__(self, project_definition: ProjectDefinition):
+        self._project = project_definition
+        self.variables = {v.name: v.value for v in project_definition.env}
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/utils/rendering.py:134]: Should we not use variables, and bundle everything under env like we discussed earlier? Could the _EnvGetter handle this instead?

File: src/snowflake/cli/plugins/nativeapp/commands.py [hunk 8]
```diff
@@ -172,7 +172,7 @@ def app_run(
 
     processor = NativeAppRunProcessor(
         project_definition=cli_context.project_definition,
-        project_root=cli_context.project_root,
+        project_root=cli_context.project_root.native_app,
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/plugins/nativeapp/commands.py:175]: did you mean to update project_definition instead?

File: src/snowflake/cli/api/commands/flags.py (hunk 1 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/commands/flags.py:488]: I don't think so, [explicite is better than implicite](https://peps.python.org/pep-0020/#the-zen-of-python). Not all command require access to project file. Imho the decorators should inform developer what is used by the command  instead of expecting them to know default values and behaviors. WDYT?

File: src/snowflake/cli/api/utils/rendering.py (hunk 5 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:104]: Should we consider using best practices for naming environment variables? For example I'm referring to
https://google.github.io/styleguide/shellguide.html#constants-and-environment-variable-names

Honestly, lower-case env vars looks strange and uncommon. I'm fine with doing a fallback to key but let's not force users to use CAPS names in snowflake.yml only to be able to write `export CAPS` instead of `export caps`. WDYT?

File: src/snowflake/cli/api/utils/rendering.py [hunk 9]
```diff
@@ -88,22 +94,151 @@ def jinja_render_from_file(
...
-        raise ValueError(f"{context_key} in user defined data")
-    context_data = {context_key: {"env": _AttrGetter(os.environ)}}
-    return {**data, **context_data}
+    if context_key in external_data:
+        raise ClickException(
+            f"{context_key} in user defined data. The `{context_key}` variable is reserved for CLI usage."
+        )
+
+    project_definition = cli_context.project_definition
+    context_data = {context_key: {"env": _EnvGetter(os.environ)}}
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:155]: Why introduce this inconsistency?

project definition present = ctx.env.foo
no pdf = env.foo

File: src/snowflake/cli/plugins/nativeapp/commands.py (hunk 8 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/plugins/nativeapp/commands.py:175]: Yes, thanks for catching this!

File: src/snowflake/cli/api/utils/rendering.py (hunk 5 above)
Comment (sfc-gh-cgorrie) [reply src/snowflake/cli/api/utils/rendering.py:104]: I'm sure CLI supports the latter on Windows (if not, I'm not sure how that ended up being the case!).

I guess the point is, why should we be opinionated about env var case sensitivity? It's already "solved" at the OS level. It's worth trying out comparables and seeing how they react to case sensitivity, as that will likely be the users' expectation. If other tooling makes things case insensitive, I'm OK with the approach here, but that isn't the case with what I'm seeing, e.g.

- Github actions uses case sensitive env vars: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable

Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:104]: > I  would guess that CLI supports the latter on Windows (if not, I'm curious how that ended up being the case!).

It's not CLI, it's the platform that supports it due to case-insensitivity. For case-sensitive platforms (which are the only concern) we expect variables to be upper case. 

Ok, let's use the keys. If someone want to follow best practices they still are free to do so using upper-case names. We just have to document this properly.

File: src/snowflake/cli/api/utils/rendering.py [hunk 10]
```diff
@@ -88,24 +96,122 @@ def jinja_render_from_file(
...
+        if var.startswith(ctx_env_prefix):
+            required_variables.append(var[len(ctx_env_prefix) :])
+    return len(found_variables), required_variables
 
 
 def snowflake_cli_jinja_render(content: str, data: Dict | None = None) -> str:
-    data = _add_project_context(data or dict())
+    data = _add_project_context(
+        external_data=data or dict(), project_definition=cli_context.project_definition
+    )
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/utils/rendering.py:216]: I think this should be part of `ProjectDefinition` validation. When we read the yaml file we should also perform the templates expansions so we don't need to worry about it in too many place. WDYT @sfc-gh-melnacouzi ?

File: tests/api/utils/test_rendering.py [hunk 11]
```diff
@@ -63,3 +70,68 @@ def test_that_undefined_variables_raise_error():
...
+    result = _add_project_context({}, project_definition=pdf)
+    assert result == {
+        "ctx": ProjectDefinition(
+            definition_version="1.1",
+            native_app=None,
+            snowpark=None,
+            streamlit=None,
+            env={"A": 42, "B": "b=42", "C": "b=42 and 42"},
+        )
+    }
```
Comment (sfc-gh-turbaszek) [inline tests/api/utils/test_rendering.py:112]: @sfc-gh-cgorrie @sfc-gh-melnacouzi if we can agree to use `key:value` format for `env` schema then this looks pretty promising. Also

File: tests/project/test_project_schemas.py [hunk 12]
```diff
@@ -63,3 +64,15 @@ def test_project_schema_is_updated_correctly_from_dict(
     native_app_project_instance.update_from_dict(update_dict)
     assert native_app_project_instance.native_app.name == "napp_test"
     assert native_app_project_instance.native_app.package.distribution == "external"
+
+
+def test_project_definition_work_for_int_version():
+    p = ProjectDefinition(definition_version=1)
+    assert p.definition_version == "1"
```
Comment (sfc-gh-turbaszek) [inline tests/project/test_project_schemas.py:71]: To assure backward compatibility

File: src/snowflake/cli/api/utils/rendering.py [hunk 13]
```diff
@@ -88,24 +96,123 @@ def jinja_render_from_file(
...
+        # Templates are reserved only to string variables
+        if not isinstance(value, str):
+            continue
+
+        # This value requires
+        include_template, required_variables = _search_for_required_variables(value)
+
+        if key in required_variables:
+            raise ClickException(
+                f"Cycle detected in environment definition for `{key}` variable."
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/utils/rendering.py:179]: We should check for `A -> B, B -> A` cycle

File: src/snowflake/cli/api/utils/rendering.py [hunk 14]
```diff
@@ -88,24 +96,123 @@ def jinja_render_from_file(
...
+            unresolved_keys.append(key)
+
+    return context_data
+
 
-    def __getattr__(self, item):
-        if item not in self._data_dict:
-            raise AttributeError(f"No attribute {item}")
-        return self._data_dict[item]
+def _check_variables_consistency(variables_data: DictWithEnvironFallback):
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/utils/rendering.py:158]: nit: rename to _get_variables_with_depedencies? The rest is just validation, so doesn't have to affect the name.

File: src/snowflake/cli/api/utils/rendering.py [hunk 15]
```diff
@@ -88,24 +96,123 @@ def jinja_render_from_file(
...
+    variables_with_dependencies = _check_variables_consistency(variables_data)
+    # Sort keys so we start from the shortest having lower probability of including more than one variable
+    unresolved_keys = sorted(list(variables_with_dependencies), reverse=True)
+
+    env = get_snowflake_cli_jinja_env()
+    context_data = {_CONTEXT_KEY: project_definition}
+    while unresolved_keys:
+        key = unresolved_keys.pop()
+        value = variables_data[key]
+        if not isinstance(value, str):
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/utils/rendering.py:144]: didn't we check this earlier?

File: src/snowflake/cli/api/utils/rendering.py [hunk 16]
```diff
@@ -88,24 +96,123 @@ def jinja_render_from_file(
...
+        DictWithEnvironFallback, project_definition.env
+    )
+
+    variables_with_dependencies = _check_variables_consistency(variables_data)
+    # Sort keys so we start from the shortest having lower probability of including more than one variable
+    unresolved_keys = sorted(list(variables_with_dependencies), reverse=True)
+
+    env = get_snowflake_cli_jinja_env()
+    context_data = {_CONTEXT_KEY: project_definition}
+    while unresolved_keys:
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/utils/rendering.py:141]: how do we guarantee no infinite loop?

File: src/snowflake/cli/api/utils/rendering.py [hunk 17]
```diff
@@ -88,24 +96,123 @@ def jinja_render_from_file(
...
+        key = unresolved_keys.pop()
+        value = variables_data[key]
+        if not isinstance(value, str):
+            continue
+
+        try:  # try to evaluate the template given current state of know variables
+            new_value = env.from_string(value).render(context_data)
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/utils/rendering.py:148]: where do environment variables play a role here? I don't see them in context_data

File: src/snowflake/cli/api/utils/rendering.py [hunk 18]
```diff
@@ -88,24 +96,123 @@ def jinja_render_from_file(
...
+            if string_includes_template(new_value):
+                unresolved_keys.append(key)
+            variables_data[key] = env.from_string(value).render(context_data)
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/utils/rendering.py:151]: why not set:
```
variables_data[key] = new_value
```
?

File: src/snowflake/cli/api/commands/flags.py [hunk 19]
```diff
@@ -515,6 +516,34 @@ def _callback(project_path: Optional[str]):
...
+                raise
+        cli_context_manager.set_project_definition(project_definition)
+        cli_context_manager.set_project_root(project_root)
+        return project_definition
+
+    return typer.Option(
+        None,
+        "-p",
+        "--project",
+        help=f"Path where Snowflake project resides. Defaults to current working directory.",
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/commands/flags.py:541]: I agree, but to limit the scope of the change and was planning to do this clean up as a followup. Will that work for you?

File: src/snowflake/cli/api/project/schemas/project_definition.py [hunk 20]
```diff
@@ -25,3 +29,26 @@ class ProjectDefinition(UpdatableModel):
...
+    env: Optional[Dict] = Field(
+        title="Environment specification for this project.",
+        default=None,
+        validation_alias="env",
+    )
+
+    @field_validator("env")
+    @classmethod
+    def _convert_env(cls, env: Optional[Dict]) -> DictWithEnvironFallback:
+        return DictWithEnvironFallback(env if env else {})
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/project/schemas/project_definition.py:41]: Yes. as suggested in other comment it would be good to move the expansion/rendering step from sql to project definition. So any use of `cli_context.project_definition` would be accessing expanded env/definition. But I decided to clean this up as a follow-up

File: src/snowflake/cli/api/project/util.py [hunk 21]
```diff
@@ -180,3 +180,7 @@ def identifier_to_show_like_pattern(identifier: str) -> str:
     matching this identifier
     """
     return f"'{escape_like_pattern(unquote_identifier(identifier))}'"
+
+
+def supported_since(version: str):
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/project/util.py:187]: Good catch, will remove!

File: src/snowflake/cli/api/utils/rendering.py [hunk 22]
```diff
@@ -88,24 +96,123 @@ def jinja_render_from_file(
...
+    """
+    if _CONTEXT_KEY in external_data:
+        raise ClickException(
+            f"{_CONTEXT_KEY} in user defined data. The `{_CONTEXT_KEY}` variable is reserved for CLI usage."
+        )
+
+    context_data = {_CONTEXT_KEY: {"env": DictWithEnvironFallback({})}}
+    # If there's project definition file then resolve variables from it
+    if project_definition and project_definition.meets_version_requirement("1.1"):
+        context_data = _resolve_variables_in_project(project_definition)
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:113]: Good idea!

File: src/snowflake/cli/api/utils/rendering.py (hunk 17 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:148]: Accessing `ctx.env` checks environment because `env` is `DictWithEnvironFallback`. So accessing any attribute will first look for environment variable.

File: src/snowflake/cli/api/utils/rendering.py (hunk 15 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:144]: True, removed and added typing

File: src/snowflake/cli/api/utils/rendering.py (hunk 10 above)
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/utils/rendering.py:217]: I agree. I will work on this as part of the project definition global expansion, since I have to do this for variables in all of the pdf.

Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/utils/rendering.py:217]: for the env part though, I think we should restrict the code from using it, or if we let code use it, we should have the DictWithEnvironFallback instance so code doesn't access what's written in env section directly.

File: src/snowflake/cli/api/commands/flags.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/commands/flags.py:541]: sounds good.

File: src/snowflake/cli/api/project/schemas/project_definition.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/project/schemas/project_definition.py:41]: makes sense. thx

File: src/snowflake/cli/api/utils/rendering.py (hunk 14 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:172]: Renamed

File: src/snowflake/cli/api/utils/rendering.py (hunk 18 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:148]: Changed 👌

File: src/snowflake/cli/api/project/schemas/project_definition.py [hunk 23]
```diff
@@ -25,3 +29,30 @@ class ProjectDefinition(UpdatableModel):
...
+        default=None,
+        validation_alias="env",
+    )
+
+    @field_validator("env")
+    @classmethod
+    def _convert_env(cls, env: Optional[Dict]) -> DictWithEnvironFallback:
+        variables = DictWithEnvironFallback(env if env else {})
+        for key in variables:
+            # Accessing value first checks for env var, so in this way we update the in-memory state
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/project/schemas/project_definition.py:43]: why is that necessary?

File: src/snowflake/cli/api/utils/rendering.py (hunk 16 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:142]: `unresolved_keys` are returned from `_get_variables_with_dependencies` that makes sure
1. None of the variables depends on unknown variable (ctx.env.something) (possible infinite loop because of unknown error)
2. There's no cycles between between variables

So `unresolved_keys` is a list of variables defined in `env` that have some cross-dependencies between `env` variables. And thanks to above resolving all of them is possible.

File: src/snowflake/cli/api/utils/rendering.py [hunk 24]
```diff
@@ -88,24 +96,122 @@ def jinja_render_from_file(
...
+    Look for &{...} pattern in  variable value. Returns a list of env variables required
+    to expand this template.`
+    """
+    ctx_env_prefix = f"{_CONTEXT_KEY}.env."
+    found_variables = re.findall(r"(&{([\.\w ]+)})+", variable_value)
+    required_variables = []
+    for _, variable in found_variables:
+        var: str = variable.strip()
+        if var.startswith(ctx_env_prefix):
+            required_variables.append(var[len(ctx_env_prefix) :])
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:226]: I can error. This will cause that `env` variables cannot reference other parts of definition file. But this can be fine for now. Do we agree to raise error and remove support for cross-project references in `env` for now?

File: src/snowflake/cli/api/utils/rendering.py [hunk 25]
```diff
@@ -88,24 +97,138 @@ def jinja_render_from_file(
...
+    for key in list(nodes):
+        q = deque([key])
+        visited: List[str] = []
+        while q:
+            curr = q.popleft()
+            if visited and curr == key:
+                raise ClickException(
+                    "Cycle detected between variables: {}".format(" -> ".join(visited))
+                )
+            visited.append(curr)
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/utils/rendering.py:180]: what's the point of visited, since we never check the content?

File: src/snowflake/cli/api/utils/rendering.py [hunk 26]
```diff
@@ -88,24 +97,138 @@ def jinja_render_from_file(
...
+
+
+def _check_for_cycles(nodes: defaultdict):
+    nodes = nodes.copy()
+    for key in list(nodes):
+        q = deque([key])
+        visited: List[str] = []
+        while q:
+            curr = q.popleft()
+            if visited and curr == key:
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/utils/rendering.py:176]: you probably need to check if curr in visited instead. What if you have a -> b -> c -> b -> c -> b ... this might cause infinite loop (I think)

File: src/snowflake/cli/api/utils/rendering.py (hunk 25 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:180]: Persisting cycle so we can construct message, I considered this useful

File: src/snowflake/cli/api/project/schemas/project_definition.py (hunk 23 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/project/schemas/project_definition.py:43]: A leftover, no longer neccessary

File: src/snowflake/cli/api/utils/rendering.py (hunk 26 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:176]: good point, fixed

**Review by sfc-gh-melnacouzi — DISMISSED** [review]

File: src/snowflake/cli/api/utils/rendering.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/utils/rendering.py:240]: yes, I agree to raise error for now. I will remove the error / tighten the check as part of my changes later.

**Review by sfc-gh-melnacouzi — DISMISSED** [review]

File: src/snowflake/cli/api/utils/rendering.py [hunk 27]
```diff
@@ -88,24 +97,152 @@ def jinja_render_from_file(
...
+
+
+def _remove_ctx_env_prefix(text: str) -> str:
+    prefix = "ctx.env."
+    if text.startswith(prefix):
+        return text[len(prefix) :]
+    return text
+
+
+def string_includes_template(text: str) -> bool:
```
Comment (sfc-gh-melnacouzi) [inline src/snowflake/cli/api/utils/rendering.py:124]: There is an issue loading yaml using the &{} template, so we will hit the chicken and egg issue here. We need to parse the yaml before we expand the templates.
Should we change the syntax - choose something parsable by yaml for project file and eventually deprecate the sql-only syntax?
Cam suggested ${}. Let me know what you think

Comment (sfc-gh-bdufour) [reply src/snowflake/cli/api/utils/rendering.py:124]: @sfc-gh-turbaszek This seems to be a critical problem with the chosen syntax. Not argument against remaining compatible with the legacy tool in SQL text, but we need to ensure that we don't end up breaking yaml parsing/generating tools because it's the _only_ syntax we support. We already know that customers like to generate a number of their project files, so that seems like an important requirement. Thoughts?

File: src/snowflake/cli/api/project/schemas/project_definition.py [hunk 28]
```diff
@@ -1,19 +1,23 @@
...
+from pydantic import Field, field_validator
 from snowflake.cli.api.project.schemas.native_app.native_app import NativeApp
 from snowflake.cli.api.project.schemas.snowpark.snowpark import Snowpark
 from snowflake.cli.api.project.schemas.streamlit.streamlit import Streamlit
 from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel
+from snowflake.cli.api.utils.models import DictWithEnvironFallback
+
+# todo: update examples
+_supported_version = ("1", "1.1")
+_latest_version = "1.1"
```
Comment (sfc-gh-pjob) [inline src/snowflake/cli/api/project/schemas/project_definition.py:15]: I don't see usage of this variable.

File: src/snowflake/cli/api/utils/models.py [hunk 29]
```diff
@@ -0,0 +1,14 @@
+import os
+
+
+class DictWithEnvironFallback(dict):
```
Comment (sfc-gh-pjob) [inline src/snowflake/cli/api/utils/models.py:4]:
```suggestion
class EnvironWithDefinedDictFallback(dict):
```

File: src/snowflake/cli/api/utils/rendering.py [hunk 30]
```diff
@@ -88,24 +97,152 @@ def jinja_render_from_file(
...
+    Updates the external data with variables from snowflake.yml definition file.
+    """
+    if _CONTEXT_KEY in external_data:
+        raise ClickException(
+            f"{_CONTEXT_KEY} in user defined data. The `{_CONTEXT_KEY}` variable is reserved for CLI usage."
+        )
+
+    context_data = _resolve_variables_in_project(project_definition)
+
+    context_data.update(external_data)
```
Comment (sfc-gh-pjob) [inline src/snowflake/cli/api/utils/rendering.py:113]: Is this correct direction of updating? Should it really be external -> context, not context -> external? IMO method comment suggests the second direction.

File: src/snowflake/cli/api/commands/flags.py (hunk 1 above)
Comment (sfc-gh-pjob) [reply src/snowflake/cli/api/commands/flags.py]: IMO it looks good :)

File: src/snowflake/cli/api/utils/rendering.py (hunk 27 above)
Comment (sfc-gh-pjob) [reply src/snowflake/cli/api/utils/rendering.py:124]: I think I agree with @sfc-gh-turbaszek that we can support using variables only as part of strings specified in quotes and cast strings properly in models if needed.

But we have to confirm that it works correctly with multiline strings which do not require quotes.

Comment (sfc-gh-bdufour) [reply src/snowflake/cli/api/utils/rendering.py:124]: We have to think long term here IMHO. We should be allowing templating in files other than the project definition file (it's on NADE's roadmap for the near future anyway). What if someone is using a numerical port number in a container spec? We can't assume strings are the only use case.

Also, we should not assume a specific parser and/or generator is used. What we make our users write should be valid YAML so that they are free to load their files in any editor, IDE or tool they choose, and it won't fail on them.

Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:124]: So the solution you propose is to use different templating for snowflake.yml. I'm fine with that.  But let's be mindful we will end up with 4 templates:
- {{ ... }} for server side
- &{ ... }  for SQL
- ${ ... }  for snowflake.yaml
- ??? for creating templates of yamls that can be rendered, this syntaxt has to be different than all previous

Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/utils/rendering.py:124]: To add on that, the idea of templating is that you replace the template with the value before reading it. In that case, if we don't use the quote, the template marker should be replaced by a non-quoted string, which is a valid YAML - which is why it doesn't seem right to force users with quoting it.

Would we consider the idea of using {{ }} instead, and have snow CLI only parse stuff that starts with ctx. ?

Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:124]: Looking for ctx requires parsing before expanding. Plus one can have SQL template in snsowlfake.yml that uses `{{ ctx }}`. Templates within templates within templates 🤯

File: src/snowflake/cli/api/utils/rendering.py (hunk 30 above)
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/utils/rendering.py:113]: I think direction doesn't matter because external_data is at a separate level in this case (top level), while the rest is under ctx.env. My understanding, external_data is only there for backward compatibility with snow sql, and will not be used for non sql stuff. @sfc-gh-turbaszek correct me if I'm wrong.

File: src/snowflake/cli/api/utils/rendering.py (hunk 27 above)
Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/utils/rendering.py:124]: I think we have to parse before expanding, and do it iteratively - at least for snowflake.yaml, because vars are defined there. Could you please elaborate on the second part about sql templates in snowflake.yml?

Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:124]: Imagine we will add support for embedding SQL in snowflake.yml. Then users may be tempted to 
```
sql:
   run: "EXECUTE IMMEDIATE  SELECT * FROM {{ ctx }}"
env:
   ...
```
In this way usage of `{{ }}` will be ambiguous. Thus, we need different template pattern.

Comment (sfc-gh-melnacouzi) [reply src/snowflake/cli/api/utils/rendering.py:124]: It's the same context, so it shouldn't matter where it's replaced. Or am I missing something?

Comment (sfc-gh-bdufour) [reply src/snowflake/cli/api/utils/rendering.py:124]: No no, I'm not proposing to use 4 different template syntax options. `{{ }}` is the server side one, so that ship has sailed. I'm proposing that we pick something for the client side that works for both SQL and YAML, and make that the client-side template. Then we also support `&{ }` as a legacy feature (maybe it's even a feature flag?). We monitor its use for deprecation purposes. If you use &{}, you use it exclusively, and only in a SQL file. Everybody else should be using the new and improved syntax that doesn't break YAML parsers. So really, it's only 2 options, other than legacy, which is something we must support regardless.

File: src/snowflake/cli/api/utils/rendering.py (hunk 30 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:113]: Correct, and in current implementation there's no option for external data to override context data (ctx key is reserved), so direction does not matter.

File: src/snowflake/cli/api/utils/rendering.py (hunk 27 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/rendering.py:124]: Having legacy for sql, still leaves us with 3 templates (legacy, jinja, something for yaml) + one more if someone wants to generate snowflake.yml, or do I miss something?

Comment (sfc-gh-bdufour) [reply src/snowflake/cli/api/utils/rendering.py:124]: Why would snowflake.yml be any different? But yes, 3 total, 2 actively supported and 1 legacy. I do think that's the minimal number. Happy to involve the product folks here if you think it would be worthwhile.

File: src/snowflake/cli/api/project/schemas/project_definition.py (hunk 28 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/project/schemas/project_definition.py:15]: Removed

File: src/snowflake/cli/api/utils/models.py (hunk 29 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/utils/models.py:4]: Renamed

**Review by sfc-gh-pjob — DISMISSED** [review]

**Review by sfc-gh-melnacouzi — APPROVED** [review]
